File: cpp/feature_service/src/kr_user_log_redis.cpp
#include "kr_user_log_redis.h"
#include <stdexcept>

inline std::string UserLogRedis::build_key(int64_t user_id) const {
    return prefix_ + ":" + std::to_string(user_id);
}

void UserLogRedis::check_reply(const redisReply* reply) {
    if (!reply) throw std::runtime_error("Redis reply null");
    if (reply->type == REDIS_REPLY_ERROR)
        throw std::runtime_error(std::string("Redis error: ") + reply->str);
}

UserLogRedis::UserLogRedis(const std::string& host, int port, int db, const std::string& key_prefix)
    : prefix_(key_prefix) {
    struct timeval timeout = {1, 500000};
    ctx_ = redisConnectWithTimeout(host.c_str(), port, timeout);
    if (!ctx_ || ctx_->err) {
        if (ctx_) {
            std::string err = "Connect: " + std::string(ctx_->errstr);
            redisFree(ctx_);
            throw std::runtime_error(err);
        } else {
            throw std::runtime_error("Connect: can't allocate ctx");
        }
    }
    auto* r = static_cast<redisReply*>(redisCommand(ctx_, "SELECT %d", db));
    check_reply(r);
    freeReplyObject(r);
}

UserLogRedis::~UserLogRedis() {
    if (ctx_) redisFree(ctx_);
}

long long UserLogRedis::append(int64_t user_id, const json& log) {
    std::string item = log.dump();
    redisReply* r = static_cast<redisReply*>(
        redisCommand(ctx_, "RPUSH %s %s", build_key(user_id).c_str(), item.c_str()));
    check_reply(r);
    long long ret = r->integer;
    freeReplyObject(r);
    return ret;
}

long long UserLogRedis::extend(int64_t user_id, const std::vector<json>& logs) {
    if (logs.empty()) return 0;
    std::vector<const char*> argv;
    std::vector<size_t> argvlen;
    argv.reserve(logs.size() + 2);
    argvlen.reserve(logs.size() + 2);
    std::string cmd = "RPUSH";
    std::string key = build_key(user_id);
    argv.emplace_back(cmd.c_str()); argvlen.emplace_back(cmd.size());
    argv.emplace_back(key.c_str()); argvlen.emplace_back(key.size());
    std::vector<std::string> items;
    items.reserve(logs.size());
    for (auto& j : logs) {
        items.emplace_back(j.dump());
        argv.emplace_back(items.back().c_str());
        argvlen.emplace_back(items.back().size());
    }
    redisReply* r = static_cast<redisReply*>(
        redisCommandArgv(ctx_, static_cast<int>(argv.size()), argv.data(), argvlen.data()));
    check_reply(r);
    long long ret = r->integer;
    freeReplyObject(r);
    return ret;
}

std::vector<json> UserLogRedis::get_slice(int64_t user_id, long long start, long long end) {
    std::vector<json> ans;
    redisReply* r = static_cast<redisReply*>(
        redisCommand(ctx_, "LRANGE %s %lld %lld", build_key(user_id).c_str(), start, end));
    check_reply(r);
    if (r->type == REDIS_REPLY_ARRAY) {
        ans.reserve(r->elements);
        for (size_t i = 0; i < r->elements; ++i)
            ans.emplace_back(json::parse(std::string(r->element[i]->str, r->element[i]->len)));
    }
    freeReplyObject(r);
    return ans;
}

std::vector<json> UserLogRedis::get_all(int64_t user_id) {
    return get_slice(user_id, 0, -1);
}

long long UserLogRedis::len(int64_t user_id) {
    redisReply* r = static_cast<redisReply*>(redisCommand(ctx_, "LLEN %s", build_key(user_id).c_str()));
    check_reply(r);
    long long ret = r->integer;
    freeReplyObject(r);
    return ret;
}

json UserLogRedis::pop_left(int64_t user_id) {
    redisReply* r = static_cast<redisReply*>(redisCommand(ctx_, "LPOP %s", build_key(user_id).c_str()));
    if (!r || r->type == REDIS_REPLY_NIL) {
        if (r) freeReplyObject(r);
        return nullptr;
    }
    json j = json::parse(std::string(r->str, r->len));
    freeReplyObject(r);
    return j;
}

json UserLogRedis::pop_right(int64_t user_id) {
    redisReply* r = static_cast<redisReply*>(redisCommand(ctx_, "RPOP %s", build_key(user_id).c_str()));
    if (!r || r->type == REDIS_REPLY_NIL) {
        if (r) freeReplyObject(r);
        return nullptr;
    }
    json j = json::parse(std::string(r->str, r->len));
    freeReplyObject(r);
    return j;
}

bool UserLogRedis::clear(int64_t user_id) {
    redisReply* r = static_cast<redisReply*>(redisCommand(ctx_, "DEL %s", build_key(user_id).c_str()));
    check_reply(r);
    bool ret = (r->integer > 0);
    freeReplyObject(r);
    return ret;
}

File: python/feature_service/kr_user_log_redis.py
#!/usr/bin/env python3
import os
import json
import sys

import redis
from typing import List, Dict, Optional
import pandas as pd
from tqdm import tqdm

class UserLogRedis:
    """
    将用户行为日志按 user_id 聚合为 Redis List（有序）
    每行 JSON 格式，字段顺序与 CSV 一致，按 time_ms 升序写入
    key -> ul:{user_id}
    """
    FIELD_NAMES = [
        "user_id", "video_id", "date", "hourmin", "time_ms",
        "is_click", "is_like", "is_follow", "is_comment", "is_forward",
        "is_hate", "long_view", "play_time_ms", "duration_ms",
        "profile_stay_time", "comment_stay_time", "is_profile_enter",
        "is_rand", "tab"
    ]

    def __init__(self, redis_client: redis.Redis, key_prefix: str = "ul"):
        self.r = redis_client
        self.prefix = key_prefix

    def _key(self, user_id: int) -> str:
        return f"{self.prefix}:{user_id}"

    def _serialize(self, row: Dict[str, str]) -> str:
        obj = {f: row[f] for f in self.FIELD_NAMES}
        return json.dumps(obj, separators=(',', ':'))

    def append(self, user_id: int, row: Dict[str, str]) -> int:
        """RPUSH 单条日志，返回当前长度"""
        return self.r.rpush(self._key(user_id), self._serialize(row))

    def extend(self, user_id: int, rows: List[Dict[str, str]]) -> int:
        """批量追加（已排好序）"""
        if not rows:
            return 0
        items = [self._serialize(r) for r in rows]
        return self.r.rpush(self._key(user_id), *items)

    def _serialize(self, row: pd.Series) -> str:
        # 保持字段顺序，直接 Series -> dict -> json
        return json.dumps(row[self.FIELD_NAMES].to_dict(), separators=(',', ':'))

    def load_csv(self, csv_path: str, batch_rows: int = 50_000) -> int:
        df = pd.read_csv(csv_path)
        grouped = df.sort_values("time_ms").groupby("user_id", sort=False)

        total = 0
        pipe = self.r.pipeline()
        for uid, grp in tqdm(grouped, desc="Loading csv"):
            # 一次性拿 5 万条
            for start in range(0, len(grp), batch_rows):
                slice_df = grp.iloc[start:start + batch_rows]
                items = slice_df.apply(self._serialize, axis=1).tolist()
                pipe.rpush(self._key(uid), *items)
                total += len(items)
            if len(pipe) >= batch_rows:
                pipe.execute()
                pipe = self.r.pipeline()
        if pipe:
            pipe.execute()
        return total

    def get_slice(self, user_id: int, start: int = 0, end: int = -1) -> List[Dict]:
        raw = self.r.lrange(self._key(user_id), start, end)
        return [json.loads(item) for item in raw]

    def get_all(self, user_id: int) -> List[Dict]:
        return self.get_slice(user_id, 0, -1)

    def len(self, user_id: int) -> int:
        return self.r.llen(self._key(user_id))

    def pop_left(self, user_id: int) -> Optional[Dict]:
        raw = self.r.lpop(self._key(user_id))
        return json.loads(raw) if raw else None

    def pop_right(self, user_id: int) -> Optional[Dict]:
        raw = self.r.rpop(self._key(user_id))
        return json.loads(raw) if raw else None

    def clear(self, user_id: int) -> bool:
        return bool(self.r.delete(self._key(user_id)))

    def clear_all(self, batch: int = 5000) -> int:
        deleted = 0
        cursor = 0
        while True:
            cursor, keys = self.r.scan(cursor, match=f"{self.prefix}:*", count=batch)
            if keys:
                deleted += self.r.delete(*keys)
            if cursor == 0:
                break
        return deleted


if __name__ == "__main__":
    ''' test redis using KuaiRand-Pure '''
    csv_path =  sys.argv[1]


    r = redis.Redis(host='localhost', port=6379, db=0, decode_responses=True)
    log = UserLogRedis(r)
    log.clear_all()
    cnt = 0
    for csv_file in ['log_standard_4_08_to_4_21_pure.csv', 'log_standard_4_22_to_5_08_pure.csv']:
        cnt += log.load_csv(os.path.join(csv_path, csv_file))
    print(f"Total users: {cnt}")

    logs = log.get_all(0)
    for i, log in enumerate(logs):
        print(f"[User 0]index{i}: {log}")